xls/modules/rle/rle.x

```diff
@@ -0,0 +1,306 @@
+//                      ──────╥───────────╥─────╥─────╥─────╥─────╥──────────
+// symbol read          XXXXXX║ A         ║ B   ║XXXXX║ B   ║ C   ║XXXXXXXXXX
+// from the symbol_r    ──────╨───────────╨─────╨─────╨─────╨─────╨──────────
+//                      ╥─────╥─────╥─────╥─────╥─────╥─────╥─────╥──────────
+// next state (symbol)  ║ 0   ║ A   ║ A   ║ B   ║ 0   ║ B   ║ C   ║ 0
+//                      ╨─────╨─────╨─────╨─────╨─────╨─────╨─────╨──────────
+//                      ╥─────╥─────╥─────╥─────╥─────╥─────╥─────╥──────────
+// next state (count)   ║ 0   ║ 1   ║ 2   ║ 1   ║ 0   ║ 1   ║ 1   ║ 0
+//                      ╨─────╨─────╨─────╨─────╨─────╨─────╨─────╨──────────
+//                                        ┌─────┐           ┌─────┐
+// next state (flush)   ──────────────────┘     └───────────┘     └──────────
+//                                        ┌───────────┐     ┌───────────┐
+// do_send              ──────────────────┘           └─────┘           └────
+//                      ──────────────────╥─────╥─────╥─────╥─────╥─────╥────
+// data send            XXXXXXXXXXXXXXXXXX║ A2  ║ B1  ║XXXXX║ B1  ║ C1  ║XXXX
+// to the data_r        ──────────────────╨─────╨─────╨─────╨─────╨─────╨────
+
+import std
+
+// structure containing output data from an RLE encoder
+struct EncOutData<SYMBOL_WIDTH: u32, COUNT_WIDTH: u32> {
+    symbol: bits[SYMBOL_WIDTH],
+    count: bits[COUNT_WIDTH],
+}
+
+// structure to preserve the state of an RLE encoder for the next iteration
+pub struct RLEEncState<SYMBOL_WIDTH: u32, COUNT_WIDTH: u32> {
+    // previous symbol and symbol count
+    prev: EncOutData<SYMBOL_WIDTH, COUNT_WIDTH>,
+    // if flush of data was requested in the previous iteration of the
+    // next function. Used to skip reading symbols from the input channel,
+    // to get time span for sending back data from the encoder.
+    flush: bool,
+
+}
+
+// RLE encoder implementation
+pub proc RLEEnc<SYMBOL_WIDTH: u32, COUNT_WIDTH: u32> {
+    // input with symbols (should be read together with flush_r)
+    symbol_r: chan<bits[SYMBOL_WIDTH]> in;
+    // input with flush request (should be read together with symbol_r)
+    flush_r: chan<bool> in;
+    // output with data, contains symbol and symbol count
+    data_s: chan<EncOutData<SYMBOL_WIDTH, COUNT_WIDTH>> out;
+
+    init {(
+        RLEEncState<SYMBOL_WIDTH, COUNT_WIDTH> {
+            prev: EncOutData<SYMBOL_WIDTH, COUNT_WIDTH> {
+                symbol: bits[SYMBOL_WIDTH]:0,
+                count: bits[COUNT_WIDTH]:0,
+            },
+            flush: false,
+        }
+    )}
+
+    config (
+        symbol_r: chan<bits[SYMBOL_WIDTH]> in,
+        flush_r: chan<bool> in,
+        data_s: chan<EncOutData<SYMBOL_WIDTH, COUNT_WIDTH>> out,
+    ) {(symbol_r, flush_r, data_s)}
+
+    next (tok: token, state: RLEEncState<SYMBOL_WIDTH, COUNT_WIDTH>) {
+        let (flush_tok, flush) = recv_if(tok, flush_r, !state.flush, false);
```

Comment:
couldn't that `recv` be also non blocking, that would allow producer not to send `flush: false` on every tick?

If we do want to enforce it, maybe we could make the `chan` type a struct instead so that we're guarantee by construct for the two to be always consistently received together (and remove the note about `(should be read together with symbol_r)`).

Reply:
fixed by using a single EncInData